apps/sms/services/js/moz_mobile_message/moz_mobile_message_shim.js

```diff
@@ -0,0 +1,205 @@
+ * @type {Array.<string>}
+ */
+const METHODS = ['getMessage', 'retrieveMMS', 'send', 'sendMMS',
+                 'delete', 'markMessageRead', 'getSegmentInfoForText'];
+
+/**
+ * Array of stream names that need to return data (messages/threads) in chunk.
+ * @type {Array.<string>}
+ */
+const STREAMS = ['getThreads', 'getMessages'];
+
+const mozMobileMessage = navigator.mozMobileMessage;
+
+var service = null;
+
+var MozMobileMessageShim = {
+  init() {
+    service = bridge.service(SERVICE_NAME);
+
+    EVENTS.forEach((event) => {
+      mozMobileMessage.addEventListener(
+        event,
+        this[Utils.camelCase(`on-${event}`)]
+      );
+    });
+
+    METHODS.forEach((shimMethod) => {
+      service.method(shimMethod, this[shimMethod]);
+    });
+
+    STREAMS.forEach((shimStream) => {
+      service.stream(shimStream, this[shimStream]);
+    });
+  },
+
+  /* Events */
+
+  onSending(e) {
+    service.broadcast('message-sending', { message: e.message });
+  },
+
+  onFailed(e) {
+    service.broadcast('message-failed-to-send', { message: e.message });
+  },
+
+  onDeliverysuccess(e) {
+    service.broadcast('message-delivered', { message: e.message });
+  },
+
+  onReadsuccess(e) {
+    service.broadcast('message-read', { message: e.message });
+  },
+
+  onSent(e) {
+    service.broadcast('message-sent', { message: e.message });
+  },
+
+  onReceived(e) {
+    service.broadcast('message-received', { message: e.message });
+  },
+
+  onDeleted(e) {
+    if (e.deletedThreadIds && e.deletedThreadIds.length) {
+      service.broadcast('threads-deleted', {
+        ids: e.deletedThreadIds
+      });
+    }
+  },
+
+  /* Methods */  
+
+  getMessage(id) {
+    return mozMobileMessage.getMessage(id);
+  },
+
+  retrieveMMS(id) {
+    return mozMobileMessage.retrieveMMS(id);
+  },
+
+  send(recipient, content, sendOpts) {
+    return mozMobileMessage.send(recipient, content, sendOpts);
+  },
+
+  sendMMS(dataOpts, sendOpts) {
+    return mozMobileMessage.sendMMS(dataOpts, sendOpts);
+  },
+
+  delete(id) {
+    return mozMobileMessage.delete(id);
+  },
+
+  markMessageRead(id, isRead, sendReadReport) {
+    return mozMobileMessage.markMessageRead(id, isRead, sendReadReport);
+  },
+
+  getSegmentInfoForText(text) {
+    return mozMobileMessage.getSegmentInfoForText(text);
+  },
+
+  /* Streams */
+
+  /**
+   * Stream wrapper for getThreads API for returning threads in chunck.
+   * @param {ServiceStream} stream channel for returning thread.
+   */
+  getThreads(stream) {
+    var cursor = null;
+
+    // WORKAROUND for bug 958738. We can remove 'try\catch' block once this bug
+    // is resolved
+    try {
+      cursor = mozMobileMessage.getThreads();
+    } catch(e) {
+      console.error('Error occurred while retrieving threads: ' + e.name);
+      stream.abort();
+      return;
+    }
+
+    cursor.onsuccess = function onsuccess() {
+      var result = this.result;
+
+      if (result) {
+        stream.write(result);
```

Comment:
I just had a quick conversation with Miller who implemented Streams about smth like
```
stream.write(result).then(() => this.continue()).catch((e) => stream.close()/abort());
```
So write will fail if stream is closed/cancelled or some unrecoverable error happened inside threads. So ideally we should track result of write somehow to stop iterating through cursor. But I think not in this PR anyway, here we just need some very basic version :)